snuba/datasets/entity.py

```diff
@@ -156,3 +174,129 @@ def describe(self) -> Description:
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        """
+        Applies entity specific validations on query argument passed
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> EntitySubscription:
+        raise NotImplementedError
+
+
+class EntitySubscriptionValidation:
+    def __init__(
+        self,
+        max_allowed_aggregations: int = 1,
+        disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"],
+        organization: Optional[int] = None,
+    ) -> None:
+        self.max_allowed_aggregations = max_allowed_aggregations
+        self.disallowed_aggregations = disallowed_aggregations
+        self.organization = organization
+
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        # Import get_entity() when used, not at import time to avoid circular imports
+        from snuba.datasets.entities.factory import get_entity
+
+        # TODO: Support composite queries with multiple entities.
+        from_clause = query.get_from_clause()
+        if not isinstance(from_clause, EntityDS):
+            raise InvalidSubscriptionError("Only simple queries are supported")
+        entity = get_entity(from_clause.key)
+
+        SubscriptionAllowedClausesValidator(
+            self.max_allowed_aggregations, self.disallowed_aggregations
+        ).validate(query)
+        if entity.required_time_column:
+            NoTimeBasedConditionValidator(entity.required_time_column).validate(query)
+
+
+class BaseEntitySubscription(EntitySubscriptionValidation, EntitySubscription):
+    """
+    This class contains all functionality required by all entities.
+    Therefore, it is currently used by all entities.
+
+    Notes:
+        If iniitialization params are not provide, it defaults to parent class attributes.
+        Setting self.organization is optional because not every entity requires this field.
+        Organization is set after Entity and EntitySubscription objects are created using set_org().
+        Child classes can extend from this to provide additional entity specific functionality.
+    """
+
+    def __init__(
+        self,
+        max_allowed_aggregations: Optional[int] = None,
+        disallowed_aggregations: Optional[Sequence[str]] = None,
+        data_dict: Mapping[str, Any] = {},
+    ) -> None:
+        super().__init__()
+        if max_allowed_aggregations and disallowed_aggregations:
+            super().__init__(max_allowed_aggregations, disallowed_aggregations)
+        self.set_org(data_dict)
+
+    def set_org(self, data_dict: Mapping[str, Any] = {}) -> BaseEntitySubscription:
```

Comment:
The existing way EntitySubscriptions work is through an EntitySubscription factory which maps an `EntityKey` to a Type[EntitySubscription]`. Therefore, the object is not created until the `SubscriptionData` is created later on and the system knows which query it wants to run. Since we are consolidating all EntitySubscription into a single class, we can't use class attributes and need to move to instance attributes because they are supposedly "different" for each entity. So with this change, when the entity is first created, it creates the `EntitySubscription` object with the appropriate parameters (max_allowed_aggregations, disallowed_aggregations). However, it isn't until later when the subscription is created that the system knows the org_id.